File: src/main/java/com/braydenoneal/blang/parser/expression/operator/ComparisonOperator.java
package com.braydenoneal.blang.parser.expression.operator;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.RunException;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.ExpressionType;
import com.braydenoneal.blang.parser.expression.ExpressionTypes;
import com.braydenoneal.blang.parser.expression.value.BooleanValue;
import com.braydenoneal.blang.parser.expression.value.FloatValue;
import com.braydenoneal.blang.parser.expression.value.IntegerValue;
import com.braydenoneal.blang.parser.expression.value.Value;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;

public record ComparisonOperator(
        String operator,
        Expression operand_a,
        Expression operand_b
) implements Operator, Expression {
    @Override
    public Value<?> evaluate(Program program) {
        Value<?> a = operand_a.evaluate(program);
        Value<?> b = operand_b.evaluate(program);

        if (operator.equals("==")) {
            return new BooleanValue(a.equals(b));
        } else if (operator.equals("!=")) {
            return new BooleanValue(!a.equals(b));
        }

        if (a instanceof IntegerValue a1 && b instanceof IntegerValue b1) {
            return switch (operator) {
                case "<=" -> new BooleanValue(a1.value() <= b1.value());
                case ">=" -> new BooleanValue(a1.value() >= b1.value());
                case "<" -> new BooleanValue(a1.value() < b1.value());
                default /* > */ -> new BooleanValue(a1.value() > b1.value());
            };
        } else if (a instanceof FloatValue a1 && b instanceof FloatValue b1) {
            return switch (operator) {
                case "<=" -> new BooleanValue(a1.value() <= b1.value());
                case ">=" -> new BooleanValue(a1.value() >= b1.value());
                case "<" -> new BooleanValue(a1.value() < b1.value());
                default /* > */ -> new BooleanValue(a1.value() > b1.value());
            };
        }

        throw new RunException("Operands are not comparable");
    }

    public static final MapCodec<ComparisonOperator> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Codec.STRING.fieldOf("operator").forGetter(ComparisonOperator::operator),
            Expression.CODEC.fieldOf("operand_a").forGetter(ComparisonOperator::operand_a),
            Expression.CODEC.fieldOf("operand_b").forGetter(ComparisonOperator::operand_b)
    ).apply(instance, ComparisonOperator::new));

    @Override
    public ExpressionType<?> getType() {
        return ExpressionTypes.COMPARISON_OPERATOR;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/operator/ComparisonOperator.kt
package com.braydenoneal.blang.parser.expression.operator

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.RunException
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.BooleanValue
import com.braydenoneal.blang.parser.expression.value.FloatValue
import com.braydenoneal.blang.parser.expression.value.IntegerValue
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder

data class ComparisonOperator(
    val operator: String,
    val operandA: Expression,
    val operandB: Expression
) : Operator, Expression {
    override fun evaluate(program: Program): Value<*> {
        val a = operandA.evaluate(program)
        val b = operandB.evaluate(program)

        if (operator == "==") {
            return BooleanValue(a == b)
        } else if (operator == "!=") {
            return BooleanValue(a != b)
        }

        if (a is IntegerValue && b is IntegerValue) {
            return when (operator) {
                "<=" -> BooleanValue(a.value <= b.value)
                ">=" -> BooleanValue(a.value >= b.value)
                "<" -> BooleanValue(a.value < b.value)
                else -> BooleanValue(a.value > b.value)
            }
        } else if (a is FloatValue && b is FloatValue) {
            return when (operator) {
                "<=" -> BooleanValue(a.value <= b.value)
                ">=" -> BooleanValue(a.value >= b.value)
                "<" -> BooleanValue(a.value < b.value)
                else -> BooleanValue(a.value > b.value)
            }
        }

        throw RunException("Operands are not comparable")
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.COMPARISON_OPERATOR

    companion object {
        val CODEC: MapCodec<ComparisonOperator> = RecordCodecBuilder.mapCodec {
            it.group(
                Codec.STRING.fieldOf("operator").forGetter(ComparisonOperator::operator),
                Expression.CODEC.fieldOf("operand_a").forGetter(ComparisonOperator::operandA),
                Expression.CODEC.fieldOf("operand_b").forGetter(ComparisonOperator::operandB)
            ).apply(it, ::ComparisonOperator)
        }
    }
}
